
$\qquad$ $\qquad$ $\qquad$  **TDA231/DIT370 Discrete Optimization: Home Assignment 3 -- LP Duality and the Primal-Dual Algorithm** <br />
$\qquad$ $\qquad$ $\qquad$                   **During grading time, direct queries regarding Q-1,2 to David & Q-3 to Marc** <br />
$\qquad$ $\qquad$ $\qquad$                     **Due Date: 03/03/2025** <br />
$\qquad$ $\qquad$ $\qquad$                   **Submitted by: Name, Personal No., Email** <br />



---


General guidelines:
*   All solutions to theoretical and pratical problems must be submitted in this ipynb notebook and equations, wherever required, should be formatted using LaTeX math-mode.
*   All discussion regarding practical problems, along with solutions and plots should be specified in this notebook. All plots/results should be visible such that the notebook does not have to be run. But the code in the notebook should reproduce the plots/results if we choose to do so.
*   Your name, personal number and email address should be specified above.
*   All tables and other additional information should be included in this notebook.
*   Before submitting, make sure that your code can run on another computer. That all plots can show on another computer including all your writing. It is good to check if your code can run here: https://colab.research.google.com.


# Problem 1.

Consider the following LP problem:

\begin{alignat*}{2}
\max \ &4x_1-2x_2+5x_3+6x_4+7x_5\\
\\
\textrm{s.t} \quad
&2x_1 + 2x_2 - 4x_3 + 4x_4 + 8x_5 &&\leq 6\\
&2x_1 + \ \ {}x_2 - 2x_3 - \ \ x_4 - 3x_5 &&\geq -1\\
&5x_1 - 2x_2 + 4x_3 + 4x_4 + 2x_5 &&= 5\\
&2x_1 - 2x_2 + 5x_3 + 3x_4 + \ \ x_5 &&\leq 4\\
&\hspace{5.3cm} \vec x &&\geq \vec 0
\end{alignat*}

* (4 points) Write the LP dual of this problem.
* (3 points) Use CVXPY to compute the primal and dual optimum solutions and compare their values.
* (3 points) Check the complementary slackness conditions.

The primal linear program above is given by:

$$
    \max  c^\top \mathbf{x} \\
    \text{s.t. } \mathbf{A}\mathbf{x} \vec{\mathcal{O}} \mathbf{b}
$$

where

<!-- \begin{bmatrix}
1 & 0 & 0 & 0 & 0\\
0 & 1 & 0 & 0 & 0\\
0 & 0 & 1 & 0 & 0\\
0 & 0 & 0 & 1 & 0 \\
0 & 0 & 0 & 0 & 1 \\
\end{bmatrix} -->

$$
\mathbf{A} =
\begin{bmatrix}
2 & 2 & -4 & 4 & 8 \\
2 & 1 & -2 & -1 & -3\\
5 & -2 & 4 & 4 & 2 \\
2 & -2 & 5 & 3 & 1 \\
\end{bmatrix}
, 
\mathbf{c} =
\begin{bmatrix}
4 \\
-2 \\
5 \\
6 \\
7 \\
\end{bmatrix}
, 
\vec{\mathcal{O}} =
\begin{bmatrix}
\leq \\
\geq \\
= \\
\leq \\
\end{bmatrix}
, 
\mathbf{b} =
\begin{bmatrix}
6 \\
-1 \\
5 \\
4 \\
\end{bmatrix}
, 
\vec x \geq \vec 0
$$


The LP dual is defined as follows:

$$
    \min b^\top \mathbf{y} \\
    \text{s.t. } \mathbf{A^\top}\mathbf{y} \vec{\mathcal{O_2}} \mathbf{c}
$$

where

$$
\mathbf{A}^T =
\begin{bmatrix} 
2 & 2 & 5 & 2 \\ 
2 & 1 & -2 & -2 \\ 
-4 & -2 & 4 & 5 \\ 
4 & -1 & 4 & 3 \\ 
8 & -3 & 2 & 1 
\end{bmatrix}
,
\vec{\mathcal{O_2}} = 
\begin{bmatrix}
\geq \\
\geq \\
\geq \\
\geq \\
\end{bmatrix}
,
\vec y
\begin{bmatrix}
\geq 0\\
\leq 0\\
\in \R \\
\geq 0 \\
\end{bmatrix}
$$

If we write it out we get the following LP:

\begin{alignat*}{2}
\min \quad & 5y_1 - y_2 + 5y_3 + 4y_4\\
\\
\textrm{s.t.} \quad
&2y_1 + 2y_2 + 5y_3 + 2y_4 &&\geq 4\\
&2y_1 + \ \ {}y_2 - 2y_3 - 2y_4 &&\geq -2\\
&-4y_1 - 2y_2 + 4y_3 + 5y_4 &&\geq 5\\
&4y_1 - y_2 + 4y_3 + 3y_4 &&\geq 6\\
&8y_1 - 3y_2 + 2y_3 + y_4 &&\geq 7\\
&y_1 \geq 0, y_2 \leq 0, y_4 \geq 0\\
\end{alignat*}


In [16]:
# Primal
import cvxpy as cp 
import numpy as np

# Variables
x = cp.Variable(5)
y = cp.Variable(4)

# Objective function
c = [4, -2, 5, 6, 7]
b = [6, -1, 5, 4]
obj_p = cp.Maximize(c @ x)
obj_d = cp.Minimize(np.transpose(b) @ y)

# Constraints
A = np.array([[2, 2, -4, 4, 8],
              [2, 1, -2, -1, -3],
              [5, -2, 4, 4, 2],
              [2, -2, 5, 3, 1]])

Ax = A @ x
A_ty = np.transpose(A) @ y

constraints_p = [Ax[0] <= b[0],
               Ax[1] >= b[1],
               Ax[2] == b[2],
               Ax[3] <= b[3],
               x >= 0]

constraints_d = [A_ty >= c,
                 y[0] >= 0,
                 y[1] <= 0,
                 y[3] >= 0]

# Solution
prob_p = cp.Problem(obj_p, constraints_p)
opt_p = prob_p.solve()

prob_d = cp.Problem(obj_d, constraints_d)
opt_d = prob_d.solve()

x_star = [float(i.value) for i in x]
y_star = [float(j.value) for j in y]

print(f"Objective value (P): {round(opt_p, 2)}")
print(f"Solution vector (P): {[round(x, 2) for x in x_star]}")

print(f"Objective value (D): {round(opt_d, 2)}")
print(f"Solution vector (D): {[round(y, 2) for y in y_star]}")

Objective value (P): 9.22
Solution vector (P): [0.63, 2.81, 1.54, 0.0, 0.66]
Objective value (D): 9.22
Solution vector (D): [0.53, -0.37, 0.34, 1.0]


The complementary slackness conditions states that:

If $(x^*, y^*)$ are a pair of optimal solutions to (P) and (D) respectively, then:

- $x^*_i = 0$, or corresponding constraint in (D) is satisfied with equality.
- $y^*_j = 0$, or corresponding constraint in (P) is satisfied with equality.

From the LP solver, we have 

$$
\mathbf{x^*} =
\begin{bmatrix}
0.63 \\
2.81 \\
1.54 \\
0 \\
0.66
\end{bmatrix}

\text{ and }

\mathbf{y^*} =
\begin{bmatrix}
0.53 \\
-0.37 \\
0.34 \\
1 \\
\end{bmatrix}
$$

We can check the slackness conditions with the following algorithm:

$$
    x^*_i = 0, \text{ or } A^Ty^*_i = c_i \text{ for all } x^*_i \in x^* \\
    y^*_j = 0, \text{ or } Ax^*_j = b_i \text{ for all } y^*_j \in y^*
$$

In [ ]:
A_Ty_star = np.transpose(A) @ y_star
Ax_star = A @ x_star
counter = 0
print("Checking for Complementary Slackness Conditions...")
for i in range(len(x_star)):
    if round(x_star[i], 2) != 0 and round(np.sum(A_Ty_star[i]), 2) != c[i]:
        counter += 1
        print(f"Complementary slackness condition not satisfied: x^*_{i} = {round(x_star[i], 2)}, A^Ty^*_{i} = {round(np.sum(A_Ty_star[i]), 2)}, c_{i} = {c[i]}")
for j in range(len(y_star)):
    if round(y_star[j], 2) != 0 and round(np.sum(Ax_star[j]), 2) != b[j]:
        counter += 1
        print(f"Complementary slackness condition not satisfied: y^*_{j} = {round(y_star[j], 2)}, Ax^*_{j} = {round(np.sum(Ax_star[j]), 2)}, b_{i} = {b[j]}")

print(f"The algorithm reached the end. {counter} violations of the Complementary Slackness Coditions found.")

Checking for Complementary Slackness Conditions ...
The algorithm reached the end. 0 violations of the Complementary Slackness Coditions found.


# Problem 2.

Consdier the LP problem:
\begin{alignat*}{2}
\max \ &6x_1 - 5x_3\\
\\
\textrm{s.t} \quad
&6x_1 - 3x_2 + x_3 &&= 2\\
&3x_1 + 4x_2 + x_3 &&\leq 5\\
&\ \ x_1 - 7x_2 &&\leq 5\\
&\hspace{2.45cm} x_1 &&\geq 0\\
&\hspace{2.45cm} x_2 &&\leq 0\\
&\hspace{2.45cm} x_3 &&\text{ unrestricted}
\end{alignat*}

* (3 points) Write the LP dual of this problem.
* (4 points) Consider the feasible solution $\vec x = (0,0,2)$ to the primal. Check if this solution is optimal by using the complementary slackness conditions to write down the corresponding dual solution.
* (3 points) Use complementary slackness to check if the primal feasible solution $\vec x = (1,0,-4)$ is optimal.

The primal LP problem above is given by:

$$  
    \max c^\top x \\
    \text{ s.t. } Ax \mathcal{O} b
$$

where

$$
\mathbf{A} =
\begin{bmatrix}
6 & -3 & 1 \\
3 & 4 & 1 \\
1 & -7 & 0 \\
\end{bmatrix}
, 
\mathbf{c} =
\begin{bmatrix}
6 \\
0 \\
-5 \\
\end{bmatrix}
, 
\vec{\mathcal{O}} =
\begin{bmatrix}
= \\
\leq \\
\leq \\
\end{bmatrix}
, 
\mathbf{b} =
\begin{bmatrix}
2 \\
5 \\
5 \\
\end{bmatrix}
, 
\vec x
\begin{bmatrix}
\geq 0 \\
\leq 0 \\
\in \R \\
\end{bmatrix}
$$

Hence, we get the LP dual:

$$
    \min b^\top \mathbf{y} \\
    \text{s.t. } \mathbf{A^\top}\mathbf{y} \mathcal{O}^\prime \mathbf{c}
$$

where

$$
\mathbf{A^\top} =
\begin{bmatrix}
6 & 3 & 1 \\
-3 & 4 & -7 \\
1 & 1 & 0 \\
\end{bmatrix}
, 
\mathcal{O}^\prime = 
\begin{bmatrix}
\geq \\
\leq \\
= \\
\end{bmatrix}
,
\vec y
\begin{bmatrix}
\in \R \\
\geq 0 \\
\geq 0 \\
\end{bmatrix}
$$

If we write it out we get the following LP:

\begin{alignat*}{2}
\min \quad & 2y_1 + 5y_2 + 5y_3\\
\\
\textrm{s.t.} \quad
&6y_1 + 3y_2 + y_3 &&\geq 6\\
&-3y_1 + 4y_2 - 7y_3 &&\leq 0\\
&y_1 + y_2 && = -5\\
&y_2 \geq 0, y_3 \geq 0\\
\end{alignat*}

We consider the feasible solution $\vec x = (0,0,2)$ to the primal (P).

The Complementary Slackness (CS) conditions states that if $\vec x$ is an optimal solution to the primal LP, then $\vec x_i = 0$, or the corresponding constraint in (D) is satisfied with equality. It is only $\vec x_3$ that does not equal 0, so the third constraint in the dual $y_1 + y_2 = -5$ must be satisfied with equality.

From $\vec x$ we have:

\begin{alignat*}{2}
&6 \cdot 0 - 3 \cdot 0 + x_3 &&= 2\\
&3 \cdot 0 + 4 \cdot 0 + x_3 &&\leq 5\\
&\ \ 0 - 7 \cdot 0 &&\leq 5\\
\end{alignat*}

The CS conditions states that if $\vec y$ is an optimal solution to the dual, then $\vec y_j = 0$, or the corresponding constraint in (P) is satisfied with equality. Since the second and third constraint in (P) is not satisfied with equality, we have that $\vec y_2 = \vec y_3 = 0$. That gives us the following constraints for (D).

\begin{alignat*}{2}
&6y_1 + 3 \cdot 0 + 0 &&\geq 6\\
&-3y_1 + 4 \cdot 0 - 7 \cdot 0 &&\leq 0\\
&y_1 + 0 && = -5\\
&0 \geq 0, 0 \geq 0\\
\end{alignat*}

The third constraint $y_1 = -5$ contradicts the two other constraints, meaning that $\vec y$ is not a feasible solution do (D). Hence $\vec x$ is not an optimal solution to (P).

<!-- $6y_1 + 3y_2 + y_3 \geq 6 \implies 6(-5 - y_2) +3y_2 + y_3 \geq 6 \implies -3y_2 + y_3 \geq 36$

$-3y_1 + 4y_2 - 7y_3 \leq 0 \implies -3(-5 - y_2) + 4y_2 - 7y_3 \leq 0 \implies 7y_2 - 7y_3 \leq 15$ -->

We consider the feasible solution $\vec x = (1,0,-4)$ to the primal (P).

From $\vec x$ we have:

\begin{alignat*}{2}
&6 \cdot 1 - 3 \cdot 0 + -4 &&= 2\\
&3 \cdot 1 + 4 \cdot 0 + -4 &&\leq 5\\
&\ \ 1 - 7 \cdot 0 &&\leq 5\\
\end{alignat*}

The second and third constraint is not satisfied with equality, and thus $y_2 = y_3 = 0$. From the constraints of (D) we get:

\begin{alignat*}{2}
&6y_1 + 3 \cdot 0 + 0 &&\geq 6\\
&-3y_1 + 4 \cdot 0 - 7 \cdot 0 &&\leq 0\\
&y_1 + 0 && = -5\\
&0 \geq 0, 0 \geq 0\\
\end{alignat*}

This gives us exactly the same result as in the previous exercise, which means that $\vec x$ is not an optimal solution to (P). 

We check the results by running the primal LP  in CVXPY.

In [3]:
import cvxpy as cp 
import numpy as np

# Variables
x = cp.Variable(3)

# Objective function
c = [6, 0, -5]
obj = cp.Maximize(c @ x)

# Constraints
A = np.array([
    [6, -3, 1],
    [3, 4, 1],
    [1, -7, 0]
])

Ax = A @ x
b = [2, 5, 5]

constraints = [Ax[0] == b[0],
               Ax[1] <= b[1], 
               Ax[2] <= b[2],
               x[0] >= 0, 
               x[1] <= 0]

# Solution
prob = cp.Problem(obj, constraints)
opt = prob.solve()
print(f"Objective value: {round(opt, 2)}")
print(f"Solution vector: {[round(float(i.value), 2) for i in x]}")

Objective value: 170.0
Solution vector: [5.0, -0.0, -28.0]


# Problem 3.

Consider the primal-dual algorithm for vertex cover discussed in class.
* (4 points) Run the algorithm by hand on the graph in the figure below (from your previous homework). Show the values of the primal and dual variables at each iteration.
* (6 points) Implement the primal-dual algorithm as a python script to compute (approximate) vertex covers and run it for the random graph $G(100,0.1)$ from the previous homework.

<img src="https://tinyurl.com/tsnuz2c" alt="Drawing" style="width: 180px;"/>

If the image does not load, try the direct link: https://tinyurl.com/tsnuz2c

We enumerate the nodes $x_a, ..., x_f$ and define the Vertex Cover problem by the following LP:

\begin{aligned}
\min \quad & 4x_a + 2x_b + 3x_c + 2x_d + x_e + 4x_f \\
\text{s.t.} \quad 
& x_a + x_b  && \geq 1 \\
& x_a + x_c  && \geq 1 \\
& x_a + x_d  && \geq 1 \\
& x_b + x_d  && \geq 1 \\
& x_c + x_d  && \geq 1 \\
& x_c + x_e  && \geq 1 \\
& x_d + x_f  && \geq 1 \\
& \vec{x} && \geq 0
\end{aligned}

From (P) we calculate (D):

\begin{aligned}
\max \quad & y_{ab} + y_{ac} + y_{ad} + y_{bd} + y_{cd} + y_{ce} + y_{df}\\
\text{s.t.} \quad 
& y_{ab} + y_{ac} + y_{ad} && \leq 4 \\
& y_{ab} + y_{bd} && \leq 2 \\
& y_{ac} + y_{cd} + y_{ce} && \leq 3 \\
& y_{ad} + y_{bd} + y_{cd} + y_{df} && \leq 2 \\
& y_{ce} && \leq 1 \\
& y_{df} && \leq 4 \\
& \vec{y} && \geq 0 
\end{aligned}

First, we initialize the solution vectors $x = 0, y = 0$.

In each iteration, we raise the value of all unfrozen $y_{ij}$ until one constraint becomes tight. Let's call that constraint the $k^{th}$ constraint in (D). Then, we freeze the edges $y_{ij}$ included in the tight constraint, and raise the corresponding vertex $x_k$ in (P). We finish the algorithm when all $y_i$ are frozen.

1) We raise the values of all $y$ values until one constraint becomes tight. This is $y_{ad} + y_{bd} + y_{cd} + y_{df} = 2$ and we freeze those vertices to $ = 0.5$. Since this is the fourth constraint, we set $x_4 = 1$.

2) We raise the values of $y_{ab} + y_{ac} + y_{ce}$ to 1 since this makes the fifth constraint tight. We freeze $y_{ce}$ and set $x_5 = 1$.

3) We raise the values of $y_{ab} + y_{ac}$ to 1.5 which makes the second and third constraint tight. We freeze both vertices and complete the algorithm.

In [40]:
# Optimal solution (ILP)

import numpy as np

# Variables
# x = [0] * 100 #  cp.Variable(100)

# # Objective function
# # obj = cp.Minimize(sum(x))

# # Constraints
# # constraints = [x >= 0]
    
# with open("graph.txt", "r") as f:
#     graph = [[int(num) for num in line.split()] for line in f]

# Test on 3a) without weights
graph = [[0, 1, 1, 1, 0, 0],
         [1, 0, 0, 1, 0, 0],
         [1, 0, 0, 1, 1, 0],
         [1, 1, 1, 0, 0, 1],
         [0, 0, 1, 0, 0, 0],
         [0, 0, 0, 1, 0, 0]]

# Create list of edges from graph
edges = []
for i in range(len(graph)):
    for j in range(len(graph[i])):
        if graph[i][j] == 1:
            edges.append((i,j))

# for (m,n) in edges:
#     constraints.append(x[m] + x[n] >= 1)


x = [0] * 6
edges = [(0,1), (0,2), (0,3), (1,3), (2,3), (2,4), (3,5)]


# Create vertex cover S
S = []

# Dual
y = [0] * len(edges) # cp.Variable(len(edges))

# For every node, a set of edges connected to it
constraints_dual = []

edge = -1
for u in range(len(graph)):
    y_e = []
    for v in range(len(graph[u])):
        if graph[u][v] == 1:
            edge += 1
            y_e.append(edge)
    constraints_dual.append(y_e)
    # constraints_dual.append(sum(y_e) <= 1)

print(f"Edges: {edges}")
print(f"Size of constraints list: {len(constraints_dual)}")
print("Constraints dual:" )
print([f"Vertex {i} : Sum of edges {constraints_dual[i]} <= 1" for i in range(len(constraints_dual))])

# Primal-dual algorithm for vertex cover

# A dictionary with frozen variables in (D)
frozen_y: dict[int, float] = {}

unfrozen_left = True
iteration = 1
while(unfrozen_left):
    print(f"Iteration {iteration}")
    # One iteration
    min = 10000 # some high number
    tight_c = -1 # to check if it is updated

    for i in range(len(constraints_dual)):
        # Raise unfrozen y/(num unfrozen y) to (1 - sum(frozen y))
        # Min = (1 - sum(frozen y)) * num unfrozen y
        num_unfrozen = 0
        c = 1
        for j in constraints_dual[i]:
            if j in frozen_y:
                c -= frozen_y[j]
            else: 
                num_unfrozen += 1
        print(f"1 - sum frozen y = {c}")
        if num_unfrozen != 0:
            min_i = c / num_unfrozen
        else: 
            min_i = min
        if min_i < min and num_unfrozen != 0:
            print(f"Min value: {min_i}")
            min = min_i
            tight_c = i

    if tight_c == -1:
        # There are no unfrozen edges left
        unfrozen_left = False
        break

    # Freeze min y's in tight constraint
    for e in constraints_dual[tight_c]:
        if e not in frozen_y:
            frozen_y.setdefault(e, min)
        print(f"Freeze edge {e}")


    # Set corresponding node to 1 in x
    # x[tight_c] = 1
    S.append(tight_c)
    print(f"Set vertex {tight_c} to 1")
    iteration += 1
    print(f"Frozen y's and their values: {frozen_y}")

print(f"Frozen variables: {dict(sorted(frozen_y.items()))}")

# Vertex cover:
print(f"Vertex cover size: {len(S)}")

vec = [0] * len(graph)
for i in S:
    vec[i] = 1
print(f"Solution vector: {vec}")


Edges: [(0, 1), (0, 2), (0, 3), (1, 3), (2, 3), (2, 4), (3, 5)]
Size of constraints list: 6
Constraints dual:
['Vertex 0 : Sum of edges [0, 1, 2] <= 1', 'Vertex 1 : Sum of edges [3, 4] <= 1', 'Vertex 2 : Sum of edges [5, 6, 7] <= 1', 'Vertex 3 : Sum of edges [8, 9, 10, 11] <= 1', 'Vertex 4 : Sum of edges [12] <= 1', 'Vertex 5 : Sum of edges [13] <= 1']
Iteration 1
1 - sum frozen y = 1
Min value: 0.3333333333333333
1 - sum frozen y = 1
1 - sum frozen y = 1
1 - sum frozen y = 1
Min value: 0.25
1 - sum frozen y = 1
1 - sum frozen y = 1
Freeze edge 8
Freeze edge 9
Freeze edge 10
Freeze edge 11
Set vertex 3 to 1
Frozen y's and their values: {8: 0.25, 9: 0.25, 10: 0.25, 11: 0.25}
Iteration 2
1 - sum frozen y = 1
Min value: 0.3333333333333333
1 - sum frozen y = 1
1 - sum frozen y = 1
1 - sum frozen y = 0.0
1 - sum frozen y = 1
1 - sum frozen y = 1
Freeze edge 0
Freeze edge 1
Freeze edge 2
Set vertex 0 to 1
Frozen y's and their values: {8: 0.25, 9: 0.25, 10: 0.25, 11: 0.25, 0: 0.33333333333333